lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -0,0 +1,149 @@
+//===- LinalgToGpu.cpp -------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TPP/Passes.h"
+
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Passes.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+using namespace mlir;
+using namespace mlir::tpp;
+
+#define GEN_PASS_CLASSES
+#include "TPP/Passes.h.inc"
+
+namespace {
+
+// Convert linalg.matmul to GPU-compatible kernel.
+struct ConvertMatmulToGpu : public OpRewritePattern<linalg::MatmulOp> {
+  using OpRewritePattern<linalg::MatmulOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    if (!matmulOp.hasBufferSemantics())
+      return rewriter.notifyMatchFailure(matmulOp,
+                                         "Linalg to GPU expects memref type");
+
+    Location loc = matmulOp.getLoc();
+    ArrayRef<int64_t> shapeC = matmulOp.getDpsInitOperand(0)
+                                   ->get()
+                                   .getType()
+                                   .cast<ShapedType>()
+                                   .getShape();
+    ArrayRef<int64_t> shapeB =
+        matmulOp.getInputs()[1].getType().cast<ShapedType>().getShape();
+    ArrayRef<int64_t> shapeA =
+        matmulOp.getInputs()[0].getType().cast<ShapedType>().getShape();
+    if (shapeB.size() == 3) {
+      return rewriter.notifyMatchFailure(matmulOp,
+                                         "Packed BF16 loops unsupported");
```

Comment:
Not all 3D matmul ops are VNNI. This needs a better matcher to make sure you get what you want here and give a reasonable error message if not.

Reply:
This check shouldn't be here at all as I go through named linalg op. To be removed.